### DISCUSSION TITLE
Result Grouping

### DISCUSSION THREAD

**Discussion by anthonycook** [meilisearch/product#531] [Feedback & Feature Proposal]
**The Problem**

Here is an example of an e-commerce search that has multiple variations of a product:

```json
{
 "name": "iPhone 14 Pro",
 "color": "Gold",
 "capacity": "128gb"
},
{
 "name": "iPhone 14 Pro",
 "color": "Space Black",
 "capacity": "128gb"
},
{
 "name": "iPhone 14 Pro",
 "color": "Deep Purple",
 "capacity": "128gb"
}
```

If I was to perform a search for "iPhone", it would return all of the results which in this example is over 16 variants, currently I believe there is no way to group these into one result.

**How Others Approach This**

In SQL we could use GROUP BY and Apache Solr archives this with FeildCollapsing and the groupValue field (see [here](https://cwiki.apache.org/confluence/display/solr/FieldCollapsingl)) 

I can think of many applications where grouping would be needed, like customer activity logs where say we want to group by email address etc.

**Comment by gmourier** [comment]
Hi @anthonycook 👋

I have some questions that may help us to know more about your needs.

What do you expect in terms of response using a group by clause?

 Is there a need to determine it on the fly at search time?

There are the `distinctAttribute` settings, but it will only return the first `iPhone 14 Pro` found if you specify `name` as the distinct attribute.

Thanks!

**Reply by anthonycook** [reply]
Hi @gmourier 👋

A response from a group clause could look something like this

```json
{
    "hits": [
        {
            "name": "iPhone 14 Pro",
            "hits": [
                {
                    "color": "Gold",
                    "capacity": "128gb"
                },
                {
                    "color": "Space Black",
                    "capacity": "128gb"
                },
                {
                    "color": "Deep Purple",
                    "capacity": "128gb"
                }
            ]
        }
    ]
}
```

An example request could be something like this

```json
{
    "q": "iPhone 14 Pro Max",
    "group": [
        "name"
    ]
}
```

I can't think of any reason why this would need to be done on the fly, maybe you could have a `groupableAttributes` field in index settings?

**Reply by gmourier** [reply]
Thanks, @anthonycook! 

We could eventually rework the `distinctAttribute` one to be named differently, accepting multiple fields in an array. At search time, you could define the number of elements to get for a group. If you specify `1`,  you would get the original behavior of the `distinctAttribute` setting.

WDYT about it?

Pinging @Kerollmops
